### PR TITLE
Add Rica integration for interactive ICA component visualization

### LINKED ISSUE #1283: Make Rica more accessible from tedana
# Feature: Web-Based Interactive Viewer for tedana Results

## Overview
Introduce a lightweight web-based viewer for tedana outputs using **Rica** (React-based interactive component analyzer). Users download a single executable binary (no dependencies required) that launches a local server and automatically opens an interactive Rica interface in their browser.

## Motivation
Current tedana workflows require users to manually inspect output files (NIfTI, JSON, images) using separate tools. A dedicated viewer would provide:
- **Unified interface** for exploring all tedana results
- **Zero setup** (single binary, no Python/Node installation)
- **Cross-platform** support (Windows, macOS, Linux)
- **Interactive visualization** (leverage modern web tech with Rica)
- **Accessibility** for non-technical users

### Architecture
Two separate repositories working together:

1. **`rica`** (Rica repository)
   - React UI for visualizing tedana results
   - Builds on every push to `main`
   - CI automatically uploads build artifacts

2. **`tedana`** (this repo)
   - Minimal Node.js server (`server.js`)
   - Bundles Rica build + server into native binaries
   - CI compiles binaries (Windows `.exe`, macOS, Linux) on release

#### User Experience
1. User downloads binary: `rica-macos` / `rica-win.exe` / `rica-linux`
2. Double-clicks to run
3. Browser automatically opens to `http://localhost:3000`
4. Rica interface loads and displays Rica

#### Technical Flow
The bundled server (`server.js`):
- **Serves Rica UI** from the `public/` folder (built HTML/CSS/JS from Rica build)
- **Provides API endpoint** `/api/files?path=/path/to/folder` that returns file listings
- **Auto-opens browser** to localhost on startup
- **Handles React Router** with catch-all route for client-side navigation

```javascript
// server.js - Minimal local server
const express = require('express');
const path = require('path');
const fs = require('fs');
const open = require('open');

const app = express();
const PORT = 3000;

// Serve Rica build
app.use(express.static(path.join(__dirname, 'public')));

// API: List files in directory
app.get('/api/files', (req, res) => {
const folderPath = req.query.path || process.cwd();
try {
const files = fs.readdirSync(folderPath, { recursive: true });
res.json(files);
} catch (err) {
res.status(500).json({ error: err.message });
}
});

// Catch-all: Serve index.html for React Router
app.get('*', (req, res) => {
res.sendFile(path.join(__dirname, 'public', 'index.html'));
});

// Start server and open browser
app.listen(PORT, () => {
console.log(Viewer running at http://localhost:${PORT});
open(http://localhost:${PORT});
});
```

#### Rica Repo Workflow (`.github/workflows/build.yml`)
- Trigger: Push to `main`
- Build Rica with `npm run build`
- Upload `build/` folder as artifact

#### tedana Repo Workflow (`.github/workflows/build-viewer.yml`)
- Trigger: Push tags (`v*`) or manual workflow dispatch
- Download latest Rica build artifact
- Bundle with `server.js` using `pkg` (Node.js binary compiler)
- Generate 3 binaries:
  - `rica-win.exe` (Windows)
  - `rica-macos` (macOS)
  - `rica-linux` (Linux)
- Attach to GitHub release for download

### Dependencies
Minimal footprint:
- **express** (~50KB): Web framework
- **open** (~30KB): Browser launcher
- **pkg**: Build tool (not bundled in final binary)

Total binary size: ~50-100MB (includes Node.js runtime)

### Phase 1: Setup
1. Ensure `rica` repository has CI workflow that builds and uploads artifacts
2. Add `server.js` and `package.json` to tedana repo
3. Configure CI/CD workflow in tedana repo to download Rica builds and compile binaries

### Phase 2: Integration
- Test cross-platform binary builds
- Verify Rica loads correctly in bundled viewer
- Distribute binaries via GitHub releases
- Document download + usage instructions

### Phase 3: Features (in Rica repo)
- Design viewer interface for tedana outputs
- Implement file browser/navigator
- Add data visualization components
- Examples to visualize:
  - Mixing matrix heatmaps
  - Component time series
  - ICA map overlays
  - Quality metrics dashboards

## Benefits

| Aspect | Benefit |
|--------|---------|
| **UX** | Single-click launch, no terminal commands |
| **Setup** | Zero dependencies (binary includes runtime) |
| **Maintenance** | Separate repos = independent development cycles |
| **Extensibility** | Easy to add new visualization features to Rica |
| **Cross-platform** | Works on all major OS without modifications |
| **Performance** | Local server = fast, no network latency |

### Security
- Server only listens on `localhost` (not exposed to network)
- File access limited to paths passed via `?path=` query parameter
- No authentication needed (local machine assumed secure)

### File Handling
- Server reads file listings recursively via `fs.readdirSync()`
- Rica app can request file contents via additional endpoints (e.g., `/api/read/<filename>`)
- Binary file parsing (NIfTI, etc.) handled client-side in Rica or via worker threads

### Scalability
- In-memory operations suitable for typical tedana outputs (~100MB typical)
- For larger datasets, implement streaming or chunked file reads

## Example Queries
Once implemented, Rica could query:
```javascript
// List all files in tedana output folder
fetch('/api/files?path=/path/to/tedana')
.then(r => r.json())
.then(files => console.log(files));

// Example response:
// ["mixing_matrix.csv", "results/component_001.nii.gz", "log.txt", ...]
```

## Alternatives Considered
1. **Web-based only** (no local app): Requires upload to server, privacy concerns
2. **Electron app**: Heavier footprint, more complex maintenance

## Related Issues/PRs
- Link to any related feature requests or discussions

## Open Questions
- What file types should Rica prioritize visualizing? (NIfTI, JSON, CSV, images?)
- Should we support drag-and-drop folder selection, or assume CLI/script usage?
- Any specific visualizations critical for your workflows?

## Acceptance Criteria
- [ ] Rica repo has CI workflow that builds and uploads artifacts
- [ ] `server.js` and `package.json` added to tedana
- [ ] CI workflow in tedana produces cross-platform binaries
- [ ] Binaries tested on Windows, macOS, Linux
- [ ] User can download and run single binary without installing dependencies
- [ ] Browser opens automatically on startup
- [ ] Rica UI successfully loads and can query `/api/files` endpoint
- [ ] Documentation added to README with download instructions

## Workflow Summary

rica repo (push to main)
↓
Build React → Upload artifact
↓
tedana repo (push tag v*)
↓
Download Rica artifact → Bundle with server.js → Compile 3 binaries
↓
Attach to GitHub release
↓
Users download → Run → Browser opens → Rica viewer ready

### PR DESCRIPTION
**Pull request by eurunuela**
## Summary

Adds Rica integration for interactive ICA component visualization. Every tedana run now creates a launcher script that handles Rica setup on demand.

## Key Changes

- **No CLI flag needed**: tedana always generates `open_rica_report.py` in the output directory
- **On-demand download**: Rica is downloaded when user runs the launcher script, not during tedana execution
- **Faster tedana runs**: No network calls during analysis
- **Self-contained launcher**: The launcher script includes all download/setup logic
- **Rica button in HTML report**: "Open in Rica" button with popup instructions

## How It Works

1. Run tedana as usual (no special flag needed)
2. Find `open_rica_report.py` in your output directory (full path is logged at end of run)
3. Run `python open_rica_report.py`
4. The script downloads Rica (if not cached), starts a local server, and opens your browser

Alternatively, open the HTML report and click the "Open in Rica" button in the top-right corner for instructions.

## Features

- Automatic Rica download from GitHub releases
- Platform-specific caching (Linux: ~/.cache, macOS: ~/Library/Caches, Windows: %LOCALAPPDATA%)
- Cross-platform path handling using `pathlib.as_posix()` for reliable Windows/Mac/Linux support
- `TEDANA_RICA_PATH` environment variable for developers to use local Rica builds
- `--force-download` flag to launcher for manual updates
- Cross-platform launcher script with built-in HTTP server and CORS support
- Graceful fallback if download fails (uses cached version if available)
- "Open in Rica" button in HTML report with modal popup

## Usage

```bash
# Run tedana
tedana -d data/*.nii.gz -e 14.5 29.0 43.5 --out-dir output

# View results in Rica (full path shown in tedana output)
python output/open_rica_report.py

# Or with ica_reclassify
ica_reclassify -c report.txt --out-dir output
python output/open_rica_report.py
```

## Test Plan

- [x] Unit tests for rica.py module (91% coverage)
- [x] Integration tests include open_rica_report.py in expected outputs
- [x] Works with both tedana and ica_reclassify workflows
- [x] Cross-platform path handling verified (Windows backslash fix)

## Files Changed

- `tedana/rica.py` - Simplified to generate self-contained launcher script
- `tedana/workflows/tedana.py` - Always calls setup_rica_report()
- `tedana/workflows/ica_reclassify.py` - Always calls setup_rica_report()
- `tedana/tests/test_rica.py` - Updated tests for new behavior
- `docs/outputs.rst` - Updated documentation
- `tedana/reporting/data/html/report_body_template.html` - Added Rica button with popup

Closes #1283.

### CODE REVIEW

**Comment by eurunuela** [comment]
## Update: Rica v2.0.0 Compatibility and Test Coverage Improvements

Just pushed commit `6fcf19a8` with the following changes:

### Changes Made:
- **Rica v2.0.0 Compatibility**: Removed `favicon.ico` from `RICA_FILES` list to support Rica v2.0.0 (file no longer provided in releases)
- **Test Coverage**: Improved from 66% to 91% (24 → 43 tests)
- **New Tests Added**:
  - Windows platform support tests (LOCALAPPDATA environment variable handling)
  - Comprehensive network function tests for `get_latest_rica_version` with API response handling
  - Asset filtering and validation tests
  - Error handling for various edge cases
  - Better test isolation with improved mocking of `is_rica_cached`

### Files Modified:
- `tedana/rica.py`: Removed `favicon.ico` from RICA_FILES list
- `tedana/tests/test_rica.py`: Added 19 new tests (+373 lines)

All pre-commit hooks passed successfully.

**Comment by tsalo** [comment]
What about adding a button to the HTML report to open the Rica website? If possible, it would be amazing to automatically select the tedana output folder as well.

**Review by handwerkerd — COMMENTED** [review]
I'm starting to look at this and noticed two issues:
1. Nearly identical messages are logged twice in a row. Not sure which location is better, but we probably want the one that includes `python`. Also, if the goal is to have this run, append the output directory onto open_rica_report.py so the command could run from the location the script was called:
```
INFO     rica:setup_rica_report:603 Rica report setup complete. Run 'open_rica_report.py' to visualize results.
INFO     tedana:tedana_workflow:1277 Rica report ready. Run 'python open_rica_report.py' to visualize results.
```

2. On my Mac, I got a crash with `python open_rica_report.py` because of 
```
open_rica_report.py", line 72
    rel_path = str(f.relative_to(cwd)).replace("\", "/")
```
It works with `replace("\\",` but I'm not sure if the double `\` might cause problems in other systems.

**Comment by eurunuela** [comment]
I have added the feedback from yesterday's meeting. Let me know what you think.

File: tedana/reporting/data/html/report_body_template.html [hunk 1]
```diff
@@ -170,7 +170,174 @@
...
+</div>
+
+<!-- Rica Modal -->
+<div class="rica-modal-overlay" id="ricaModal">
+  <div class="rica-modal">
+    <button type="button" class="rica-modal-close" onclick="closeRicaModal()">×</button>
+    <h2>Open in Rica</h2>
+    <p>
+      <strong>Rica</strong> is an interactive web-based tool for exploring ICA components
+      with 3D brain visualization.
```
Comment (handwerkerd) [inline tedana/reporting/data/html/report_body_template.html]:
```suggestion
      with 3D brain visualization and
      can be used to manually change component classifications.
```

**Review by handwerkerd — CHANGES_REQUESTED** [review]
The changes all look good.
I did notice one issue. The bugs in rica visualization that you fixed appeared in rica when I opened it with `python open_rica_report.py`. The same report rendered properly when I opened it with my local version of rica using `npm install; npm start` I'm fairly sure what happened is that, when the program checks to see if it should download rica, it saw it was already present and didn't re-download. You might want some way to see if there's a newer version of rica available & to then download that version.

**Comment by eurunuela** [comment]
Ah good catch! I had the latest on mine and didn't see this. Thanks Dan!

**Comment by handwerkerd** [comment]
It still didn't work, but I noticed you mentioned browser caching so I opened `http://localhost:8000/rica/index.html` in a different browser and it works. It does seem like the issue is the older version is cached in the browser.

**Comment by eurunuela** [comment]
I think this was only an issue for you because you already had the previous cache. I don't think it should be an issue for users that start to use it once it's merged. I think the cache clearing will work for them. In any case, I added this to both tedana's (briefly) and rica's documentation.

**Review by handwerkerd — APPROVED** [review]
Glad to see this added in!

**Comment by eurunuela** [comment]
Do you think we need a second reviewer for this? Or are we good with one?
